Support non-destructive sets in conditional sections - RFC

evs-chris opened the pull request:
## Description of the pull request:

When a rendered conditional section's condition goes false-y, the section is completely torn down and unrendered. For tiny chunks of template, that doesn't mean very much, but for larger things that means that the teardown is expensive and setting it back up when it goes truth-y again is very expensive. It also means that any components within that section are entirely destroyed and recreated in the same cycle, which can have unfortunate side-effects for components with private internal state.

This is my attempt to address those issues by adding a flag to the mutation methods `set`, `add`, `subtract`, and `toggle` that tells any sections that unrender during the course of updates propagating from the model mutation not to completely destroy their fragment but, instead, just unrender it and hang on to it for future re-rendering. The flag is passed in a new options object as the last parameter of those methods e.g. `ractive.toggle('foo', { keep: true })`. The default behavior is still to tear everything down.

### TODO

* [x] Tests

I haven't done any extensive testing yet, but this theoretically supports transitions properly. I also haven't specifically tested it with components, but I have verified that DOM is not completely torn down and recreated when `keep === true`.

## Fixes the following issues:
Just aggravation I've experienced since trying to write a tab component that used conditional sections before giving up and resorting to `display: none` instead, which is not at all transition friendly.

## Is breaking:
Nope.

## Reviewers:
Yep.

evs-chris committed:
experimental support for keeping section fragments around during unrender when they would normally be discarded

ghost commented:
Sounds like another one of those "potentially an optimization" things or just do manual hiding.

In the case of optimization, what's the downside of hanging onto it by default? If I remember correctly, I read somewhere that Inferno does this to make it real fast, where it doesn't completely destroy stuff but instead holds on to them for future use.

But then, one can simply just do manual hiding using CSS if one really wants to preserve the DOM. In the case of tabs, it's simply a `ractive.set` toggling the some value that controls the presence of a class on each tab and tab content. If I need animation, I jujst pop in CSS transitions with numeric properties like `max-width`, `max-height` and `opacity` in place of `display:none`.

evs-chris committed:
account for set with an object and options

evs-chris commented:
There are pros and cons to both keeping and throwing away DOM and vDOM. If you keep it and you actually don't want it any more, then that's a bit more garbage sitting around that can't be reclaimed. It can add up pretty quickly with complex apps, and ractive is already at a bit of a disadvantage there because the model tree is never pruned. This is probably a bigger deal on mobile. For reference, my tests show a ~75%/25% improvement on render/unrender when `keep: true` is used.

As for manual CSS transitions, that's how I've handled it too, but it leaves a bit to be desired. Ractive intro transitions have the tremendous benefit of being able to measure exactly how big an element needs to be when the transition ends, which makes scaling/sliding/fly type transitions much more pleasant. Getting a transition sequence to end with a `display: none` and start with `display: block` (or whatever) also requires a bit of timing and class voodoo.

evs-chris committed:
tests for set keep: true, which now works with triples and fully reuses existing nodes
also got carried away and added progressive enhancement for triples and text node splitting

evs-chris committed:
use for loop rather than foreach in fragment renders since they're hot and it makes profiling much easier

evs-chris commented:
There are now tests, and in adding them, I discovered that nodes were being recreated rather than reused. That is now fixed.

I also got a bit carried away and did some progressive enhancement fixes while updating kept fragment support. Triples will now reuse nodes during initial render with `enhance: true`. Text nodes and interpolators will use `Text.splitText()` rather than overwriting the initial node and adding new ones. I'm gonna call that close enough for now, so this will now fix #2664 and #2403.

evs-chris committed:
add a few more steps to the jsweb tests and more data points to the output table for comparison
